Generator/Sources/NeedleFramework/Entry/Generator.swift

```diff
@@ -109,9 +128,92 @@ public class Generator {
+    /// - parameter retryParsingOnTimeoutLimit: The maximum number of times
+    /// parsing Swift source files should be retried in case of timeouts.
+    /// - parameter concurrencyLimit: The maximum number of tasks to execute
+    /// concurrently. `nil` if no limit is set.
+    /// - throws: `GeneratorError`.
+    public final func printDependencyTree(from sourceRootPaths: [String],
+                             withSourcesListFormat sourcesListFormatValue: String? = nil, 
+                             excludingFilesEndingWith exclusionSuffixes: [String], 
+                             excludingFilesWithPaths exclusionPaths: [String],
+                             shouldCollectParsingInfo: Bool, 
+                             parsingTimeout: Double, 
+                             retryParsingOnTimeoutLimit: Int, 
+                             concurrencyLimit: Int?,
+                             rootComponentName: String) throws {
+        let sourceRootUrls = sourceRootPaths.map { (path: String) -> URL in
+            URL(path: path)
+        }
+        
+        let executor = createExecutor(withName: "Needle.generate", shouldTrackTaskId: shouldCollectParsingInfo, concurrencyLimit: concurrencyLimit)
+
+        var retryParsingCount = 0
+        while true {
+            do {
+                let parser = DependencyGraphParser()
+                let (components, _) = try parser.parse(from: sourceRootUrls, withSourcesListFormat: sourcesListFormatValue, excludingFilesEndingWith: exclusionSuffixes, excludingFilesWithPaths: exclusionPaths, using: executor, withTimeout: parsingTimeout)
+                let printer = DependencyGraphPrinter(components: components)
+                printer.printDependencyTree(rootComponentName: rootComponentName)
+                break
+            } catch DependencyGraphParserError.timeout(let sourcePath, let taskId) {
+                retryParsingCount += 1
+                let message = "Parsing Swift source file at \(sourcePath) timed out when executing task with ID \(taskId). SourceKit daemon process status: \(sourceKitUtilities.isSourceKitRunning)."
+                if retryParsingCount >= retryParsingOnTimeoutLimit {
+                    throw GeneratorError.withMessage(message)
+                } else {
+                    warning(message)
+                    warning("Attempt to retry parsing by killing SourceKitService.")
+                    // Killing the SourceKit process instead of issuing a SourceKit shutdown command, since the SourceKit
+                    // process is likely hung at this point.
+                    let didKill = sourceKitUtilities.killProcess()
+                    if !didKill {
+                        warning("Failed to kill SourceKitService.")
+                    }
+                    sourceKitUtilities.initialize()
+                }
+            } catch {
+                throw GeneratorError.withMessage("Unknown error \(error).")
+            }
+        }
+    }
```

Comment:
Seems like this is copied over code from the `generate` command. Can we create a common method and pass in a closure or some sort?